### PR TITLE
Distributed merge sort

### PR DESCRIPTION
**Pull request by arhimondr**
The MergeOperator allow sorting to be distributed across the cluster via a distributed merge sort.

The merge operator merges streams of data from N remote tasks, retrieving the buffers just like exchange operators do. The input data must be sorted, and in order to stream output, the merge operator must have a buffer from each remote task (or that remote task must have no more data).

The plan for `select * from nation order by nationkey` is:
```
                                                                                Query Plan                                                              
-------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[nationkey, name, regionkey, comment] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                         
         Cost: {rows: ?, bytes: ?}                                                                                                                     
     - RemoteMerge[nationkey ASC_NULLS_LAST] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                           
             Cost: {rows: ?, bytes: ?}                                                                                                                 
         - LocalMerge[nationkey ASC_NULLS_LAST] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                        
                 Cost: {rows: ?, bytes: ?}                                                                                                             
             - PartialSort[nationkey ASC_NULLS_LAST] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]                   
                     Cost: {rows: ?, bytes: ?}                                                                                                         
                 - LocalExchange[ROUND_ROBIN] () => nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)                         
                         Cost: {rows: 25, bytes: ?}                                                                                                    
                     - RemoteExchange[ROUND_ROBIN] => nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)                       
                             Cost: {rows: 25, bytes: ?}                                                                                                
                         - TableScan[tpch:tpch:nation:sf0.01, originalConstraint = true] => [nationkey:bigint, name:varchar(25), regionkey:bigint, comm
                                 Cost: {rows: 25, bytes: ?}                                                                                            
                                 nationkey := tpch:nationkey                                                                                           
                                 name := tpch:name                                                                                                     
                                 regionkey := tpch:regionkey                                                                                           
                                 comment := tpch:comment     
```

Preliminary redistribution before the partial sort is configurable and can be disabled on session level.

### CODE REVIEW

**Comment by kokosing** [comment]
@martint Please hold on with review of this for a while, until some one from @sopel39 @findepi @losipiuk or me give accept this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java [hunk 1]
```diff
@@ -62,14 +64,17 @@
     // for each source, the list of inputs corresponding to each output
     private final List<List<Symbol>> inputs;
 
+    private final Optional<OrderingScheme> orderingScheme;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: I would extract this change as separate commit, it follows with several mechanical changes related to constructor invocations, so it would be nice to have it separated

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MergeRemoteSourceNode.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+@Immutable
+public class MergeRemoteSourceNode
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MergeRemoteSourceNode.java]: I would extract adding this node as separate commit

File: presto-tests/src/test/java/com/facebook/presto/tests/TestDistributedSortQueriesNoRedistribute.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            super(".*OrderBy.*");
+        }
+    }
+
+    public TestDistributedSortQueriesNoRedistribute()
+            throws Exception
+    {
+        super(() -> TpchQueryRunner.createQueryRunner(ImmutableMap.of(
+                "experimental.distributed-sort", "true",
+                "experimental.redistribute-sort", "false")));
```
Comment (kokosing) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestDistributedSortQueriesNoRedistribute.java]: I thought that false is default, should not be here true?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestDistributedSortQueriesNoRedistribute.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.facebook.presto.tests.tpch.TpchQueryRunner;
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Listeners;
+
+@Listeners(TestDistributedSortQueriesNoRedistribute.MethodSelector.class)
+public class TestDistributedSortQueriesNoRedistribute
+        extends AbstractTestQueries
+{
+    public static class MethodSelector
```
Comment (kokosing) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestDistributedSortQueriesNoRedistribute.java]: Maybe extract this as separate class and reuse in both ATQ implementations

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 5]
```diff
@@ -122,7 +122,6 @@
 import com.facebook.presto.sql.planner.plan.JoinNode;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: please squash this commit

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java [hunk 6]
```diff
@@ -144,21 +168,28 @@ public PageWithPosition getPage()
...
         private int position = 0;
 
-        public PageWithPosition(Page page)
+        public PageWithPosition(int streamIndex, Page page)
         {
+            this.streamIndex = streamIndex;
             this.page = requireNonNull(page, "page is null");
         }
 
+        public int getStreamIndex()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: should not be this private?

**Comment by kokosing** [comment]
@findepi Please take a look

File: presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class OrderingScheme
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java]: Once you do this, I think it would be nice to extract this as separate PR and merge this faster as this is nice refactor. For example PR with ORDER BY in aggregation might take advantage of it.

File: presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java [hunk 8]
```diff
@@ -31,6 +31,7 @@ public static void assertPageEquals(List<? extends Type> types, Page actualPage,
     {
         assertEquals(types.size(), actualPage.getChannelCount());
         assertEquals(actualPage.getChannelCount(), expectedPage.getChannelCount());
+        assertEquals(actualPage.getPositionCount(), expectedPage.getPositionCount());
```
Comment (cawallin) [inline presto-main/src/test/java/com/facebook/presto/operator/PageAssertions.java]: https://github.com/prestodb/presto/pull/8354

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: but it is usage from the same file so it can be private

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 9]
```diff
@@ -73,6 +73,7 @@
     public static final String ENABLE_INTERMEDIATE_AGGREGATIONS = "enable_intermediate_aggregations";
     public static final String PUSH_AGGREGATION_THROUGH_JOIN = "push_aggregation_through_join";
     public static final String PUSH_PARTIAL_AGGREGATION_THROUGH_JOIN = "push_partial_aggregation_through_join";
+    public static final String DISTRIBUTED_SORT = "distributed_sort";
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: You can make it disabled by default in the commit when it is added.

**Comment by kokosing** [comment]
Please squash the fixup commits

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static java.util.Objects.requireNonNull;
+
+public class MergeSortProcessor
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: Have you considered reusing `com.facebook.presto.operator.MergeHashSort`? It seems to be providing similar functionality and a lot of concepts are repeating (e.g: `PageWithPosition`). `MergeHashSort` currently uses hash value for sorting, but it should be easy to extend it so it takes `MergeSortComparator/PageComparator` and possibly `PageSuppliers` instead of `Iterators<Page>`

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortComparator.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+
+public interface MergeSortComparator
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortComparator.java]: I would rename it to `PageComparator`.

File: presto-main/src/main/java/com/facebook/presto/operator/PageSupplier.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+
+public interface PageSupplier
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/PageSupplier.java]: Maybe this should be a more generic concept of iterator, for instance:
```
public interface BlockingIterator<T> extends Iterator<T> {
  // equivalent of isFinished (false when no more elements are expected)
  boolean hasNext();
  // returns next element, might block
  @Nullable
  T next();
  ListenableFuture<?> isBlocked();
}
```
It would be then possible to cleanly extend `MergeHashSort` and use it in distributed sort code. Also the abstraction could be used in other places, e.g: `MergeSource` and `MergeSortProcessor` could implement `BlockingIterator<PageWithPosition>` and `BlockingIterator<Page>`. One could even imagine standard transformation methods and base classes for such iterators (similarly as for futures and Guava `AbstractIterator`). There could be a utility method that would decorate existing iterator with blocking future.

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java (hunk 10 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: This class doesn't account for memory usage, see how `MergeHashSort` is doing that. That is why unification of both classes would be beneficial.

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+        blocked = Futures.allAsList(mergeSources.stream().map(MergeSource::isBlocked).collect(Collectors.toList()));
+        return blocked;
+    }
+
+    public boolean isFinished()
+    {
+        return mergeSources.stream().allMatch(MergeSource::isFinished);
+    }
+
+    private static class MergeSource
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: It doesn't account for memory, see: `com.facebook.presto.operator.MergeHashSort.SingleChannelPagePositions`

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java (hunk 10 above)
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: The `ExchangeClient` itself accounts for memory usage when it accepts pages (see `ExchangeClient:addPages`, line 227); we wouldn't want to double count it. I see how it is pretty similar functionality, but considering the fact that memory would be accounted differently and this would be an invasive change to spill for aggregations, is it worth it?

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+        requireNonNull(pageSuppliers, "pageSuppliers is null");
+        ImmutableList.Builder<MergeSource> mergeSources = ImmutableList.builder();
+        for (PageSupplier pageSupplier : pageSuppliers) {
+            mergeSources.add(new MergeSource(pageSupplier));
+        }
+        this.mergeSources = mergeSources.build();
+    }
+
+    @Nullable
+    public PageWithPosition poll()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: I like the idea of `appendTo` method. It is very coherent with other classes which are doing similar thing.

File: presto-main/src/test/java/com/facebook/presto/operator/TestMergeSortProcessor.java [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+            executor = null;
+        }
+    }
+
+    @Test
+    public void testSingleStream()
+            throws Exception
+    {
+        List<Type> types = ImmutableList.of(INTEGER, INTEGER);
+        MaterializedResult actual = mergeSortedPages(
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/operator/TestMergeSortProcessor.java]: Lets make this decision in runtime. Before you can start merge you need to collect all the input streams (splits). So let create a factory that is collecting splits and once it collect them all it is able to create a `MergeSortOperator`, then in case of single stream it would just return NO-OP kind of MSO, otherwise the actual implementation.

Such kind of approach does not complicate much.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java (hunk 7 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java]: I see your point, but ;)

 - `ExchangeNode` is nothing special, is just yet another plan node and now in case of having `OrderingScheme` it ensures ordering (exactly like SortNode)
 - No worries, we can always do this kind of refactor as follow up PR, let keep as it is for a while

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: Thank you, IMO it is better ;)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java [hunk 16]
```diff
@@ -62,14 +64,17 @@
...
+
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: The more I think about it, the less convinced I become that this `orderingScheme` should belong to `ExchangeNode`. IMO it makes `ExchangeNode` (and all the related operators) to be somehow overwhelmed with responsibilities.

Is not that `ExchangeNode` should preserve the order of rows (FIFO)? If so why not to sort before the exchange. That way you would not need change this class at all. What do you think?

Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: Yeah, it is weird. However, `ExchangeNode` and `ExchangeOperator` definitely do not preserve order -- they just return the pages as they come from all of the remote sources. We use this `ExchangeNode` to create a remote source that knows about ordering so you can merge. We do sort before the exchange, but you additionally need to merge the inputs as they come in. It's best to do this in the exchange because you need to be able to map pages to where they're coming from (for the merge sort algorithm -- since you need to have a page from all of the sources so you are sure that you are taking the smallest value).

Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: What we had before was a separate PlanNode for merging (MergeExchangeNode). However, ExchangeNode and MergeExchangeNode should behave pretty much the same for all of the optimizer rules, so you have a ton of duplication and opportunity for error if you introduce a new node.

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java (hunk 14 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: > I agree that it's weird, but is it worth a data copy to make the code cleaner?

You don't need to copy the data, just the create new `PageWithPosition` with incremented position variable. Sure, you would need to create new `PageWithPosition` for each row, but I think it's worth it for the sake of simplicity and cleanness. 

See how `MergeHashSort` is doing that with iterator approach: `MergeHashSort.SingleChannelPagePositions` and `MergeHashSort.PageRewriteIterator`. It uses Guavas `Iterators.mergeSorted` to have log number of comparisons instead of linear

File: presto-main/src/test/java/com/facebook/presto/operator/TestMergeSortProcessor.java (hunk 15 above)
Comment (cawallin) [inline presto-main/src/test/java/com/facebook/presto/operator/TestMergeSortProcessor.java]: You need to create the operator to start collecting splits, but we could add logic that doesn't use the `MergeSortProcessor` when you have only one split.

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java (hunk 14 above)
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: @sopel39 was suggesting merging `MergeSortProcessor` with `MergeHashSort` -- I'll have to think about what that'll look like, including this comment about `appendTo`. If you have any opinions/suggestions, let me know.

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+    {
+        PageWithPosition result = null;
+        for (MergeSource mergeSource : mergeSources) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: this has linear number of comparisons. You could do log number of comparisons. `MergeHashSort` does that with iterators and  Guavas `Iterators.mergeSorted`

File: presto-main/src/main/java/com/facebook/presto/operator/PageSupplier.java (hunk 12 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/PageSupplier.java]: I think some kind of `Iterators#mergeSorted` that supports `BlockingIterator<T>` would need to be implemented.

Alternatively, we could have a special class that stores either value or blocked future:
```
class ValueOrBlocked<T> {
  ...
  Optional<T> getValue();
  Future<?> isBlocked();
}
```
and use it in iterators:
```
// page supplier
Iterator<ValueOrBlocked<Page>> ...
// page with position supplier
Iterator<ValueOrBlocked<PageWithPosition>> ...

// use it in merging with Guava (blocked ValueOrBlocked have precedence over non-blocked values)
Iterator<ValueOrBlocked<PageWithPosition>> Iterators.mergeSorted(pageWithPositionIterators);
```
think this approach is more hacky, but allows us to reuse Guava `mergeSorted`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java [hunk 18]
```diff
@@ -91,11 +97,14 @@ public ExchangeNode(
 
         checkArgument(scope != REMOTE || type == Type.REPARTITION || !partitioningScheme.isReplicateNullsAndAny(), "Only REPARTITION can replicate remotely");
 
+        checkArgument(type == Type.GATHER || !orderingScheme.isPresent(), "ordering scheme can be set only for GATHERING");
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: Do you know that `PartitioningScheme` allows for `orderingScheme` to be present when `type=GATHER`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: Don't you think that merging deserve separate `ExchangeNode.Type=MERGE`? This way we can clearly separate simple gathering from more complex merging. In the future we could use this type to do intermittent merging. What do you think? I don't have strong opinion about this.

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: Currently ordering symbols are part of output layout and are materialized. Please add a TODO to not materialize ordering symbols when not needed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java [hunk 19]
```diff
@@ -13,6 +13,7 @@
  */
 package com.facebook.presto.sql.planner.plan;
 
+import com.facebook.presto.sql.planner.OrderingScheme;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: please add proper `OrderingSchema` formatting support to `PlanPrinter.Visitor#visitExchange`

File: presto-main/src/test/java/com/facebook/presto/operator/OperatorAssertion.java [hunk 20]
```diff
@@ -217,6 +227,50 @@ public static void assertOperatorEqualsIgnoreOrder(
         assertEqualsIgnoreOrder(actual.getMaterializedRows(), expected.getMaterializedRows());
     }
 
+    public static void assertOperatorIsBlocked(Operator operator)
+    {
+        assertOperatorIsBlocked(operator, BLOCKED_DEFAULT_TIMEOUT);
+    }
+
+    public static void assertOperatorIsBlocked(Operator operator, Duration timeout)
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/operator/OperatorAssertion.java]: I would prefer active waiting here (e.g with 10ms intervals) for operator to become blocked. This way we don't have to wait whole 100 millis and also make test more stable since 100ms seems like a short time

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java (hunk 6 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java]: We had similar discussion before. Outside the class encapsulation is ensured by making the inner class to be private. Making this method public has no effect from the compiler perspective,  however it suggest what methods are being used from the outer class, and what methods are used by the inner class only.

File: presto-main/src/main/java/com/facebook/presto/operator/MergeOperator.java [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+    public Page getOutput()
+    {
+        checkState(mergeSortProcessor != null, "mergeSources must be non-null when getting output");
+
+        if (closed) {
+            return null;
+        }
+
+        while (!pageBuilder.isFull()) {
+            PageWithPosition pageWithPosition = mergeSortProcessor.poll();
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeOperator.java]: yes

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java (hunk 1 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java]: Then in many places in optimizer you need to handle `MERGE` type the same way you handle `GATHER`

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortComparator.java (hunk 11 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortComparator.java]: But it's not comparator of pages. `PagePositionComparator`? `ComparisonFunction`?

File: presto-main/src/main/java/com/facebook/presto/operator/MergeSortComparator.java [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+{
+    int compareTo(Page left, int leftPosition, Page right, int rightPosition);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/MergeSortComparator.java]: You could resist temptation to mimic `Comparator` and use more meaningful return type here (an enum)... (I didn't read other code, so it may not make any sense though)

File: presto-main/src/test/java/com/facebook/presto/operator/OperatorAssertion.java [hunk 23]
```diff
@@ -217,6 +227,48 @@ public static void assertOperatorEqualsIgnoreOrder(
...
+        try {
+            operator.isBlocked().get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+            return true;
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("interrupted", e);
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e.getCause());
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/OperatorAssertion.java]: `throw new RuntimeException(e);` would be better

**Comment by cawallin** [comment]
@losipiuk @kokosing @sopel39 all of the comments have been answered -- if you're still interested in reviewing, please do so, else we'll pass it on to others to review.

**Comment by kokosing** [comment]
I am sorry, but I am afraid I won't be able to do decent review of this PR in the near future.

**Comment by cawallin** [comment]
@dain @martint could you take a look, or ask someone else to? This patch has finished TD review.

**Comment by martint** [comment]
@cawallin, yes, I'll put it in my queue.

**Comment by sopel39** [comment]
Superseded by: https://github.com/prestodb/presto/pull/9854
